File: app/src/test/java/com/example/adiphong/rx2/FromCallableTest.java
package com.example.adiphong.rx2;

import org.junit.Test;

import java.util.Arrays;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;

/**
 * Created by adiphong on 27/11/2017 AD.
 */

public class FromCallableTest {

// String
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    @Test
    public void flowableFromCallable() {
        Flowable.fromCallable(
                () -> "hello")
                .subscribe(
                        s -> System.out.println("------- flowableFromCallable " + s)
                );
    }

    @Test
    public void observableFromCallable() {
        Observable.fromCallable(
                () -> "hello")
                .subscribe(
                        s -> System.out.println("------- ObservableFromCallable " + s)
                );
    }

    @Test
    public void maybeFromCallable() {
        Maybe.fromCallable(
                () -> "hello")
                .subscribe(
                        s -> System.out.println("------- MaybeFromCallable " + s)
                );
    }

    @Test
    public void singleFromCallable() {
        Single.fromCallable(
                () -> "hello")
                .subscribe(
                        s -> System.out.println("------- SingleFromCallable " + s)
                );
    }

    @Test
    public void completableFromCallable() {
        // ignore result
        Completable.fromCallable(
                () -> "hello")
                .subscribe(() -> System.out.println("------- completableFromCallable ignore result")
                );
    }

    // Integer
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Test
    public void intObservableFromCallable() {
        Observable.fromCallable(
                () -> {
                    int test = 0;
                    return test / 0;
                })
                .subscribe(
                        s -> System.out.println("------- intObservableFromCallable " + s),
                        throwable -> System.out.println("------- intObservableFromCallable onError ")
                );
    }


    // List
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Test
    public void listObservableFromCallable() {
        Observable.fromCallable(
                () -> Arrays.asList("hello", "world"))
                .subscribe(
                        s -> System.out.println("------- listObservableFromCallable " + s)
                );
    }
}

File: app/src/test/java/com/example/adiphong/rx2/FromCallableTest.kt
package com.example.adiphong.rx2

import io.reactivex.*
import org.junit.Test

/**
 * Created by adiphong on 27/11/2017 AD.
 */

class FromCallableTest {

    // String
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    @Test
    fun flowableFromCallable() {
        Flowable.fromCallable {
            "hello"
        }.subscribe { s ->
            println("------- flowableFromCallable $s")
        }
    }

    @Test
    fun observableFromCallable() {
        Observable.fromCallable {
            "hello"
        }.subscribe { s ->
            println("------- ObservableFromCallable $s")
        }
    }

    @Test
    fun maybeFromCallable() {
        Maybe.fromCallable {
            "hello"
        }.subscribe { s ->
            println("------- MaybeFromCallable $s")
        }
    }

    @Test
    fun singleFromCallable() {
        Single.fromCallable {
            "hello"
        }.subscribe { s ->
            println("------- SingleFromCallable $s")
        }
    }

    @Test
    fun completableFromCallable() {
        // ignore result
        Completable.fromCallable {
            "hello"
        }.subscribe {
            println("------- completableFromCallable ignore result")
        }
    }

    // Integer
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Test
    fun intObservableFromCallable() {
        Observable.fromCallable {
            val test = 0
            test / 0
        }.subscribe(
                { s -> println("------- intObservableFromCallable " + s!!) },
                { throwable -> println("------- intObservableFromCallable onError ") }
        )
    }

    // List
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    @Test
    fun listObservableFromCallable() {
        Observable.fromCallable {
            listOf("hello", "world")
        }.subscribe { s ->
            println("------- listObservableFromCallable $s")
        }
    }
}
